### ISSUE TITLE
Check for zeros in the data

### ISSUE THREAD

**Issue by peterdudfield** [openclimatefix/nwp-consumer#159]
## Detailed Description
It would be great to have a check in place that checks for zeros. A large amount of these is normally an error

## Context
- good to catch data problems early and fail hard

## Possible Implementation
- add a a check, if number os zeros >20% of the data, error
- @devsjc can you suggest a place in the code for this check?

**Comment by devsjc** [comment]
Yes, I would add it to the filter here: https://github.com/openclimatefix/nwp-consumer/blob/372359ce7789a8da62964bbc333484d2343ee8fa/src/nwp_consumer/internal/service/consumer.py#L410-L424

**Comment by GAMinsect** [comment]
Hi, I'm someone who is just starting in the field of open source development. 
I would like to know if i can contribute to this, since it' a good first issue.
Thank you in advance.

**Comment by peterdudfield** [comment]
> Hi, I'm someone who is just starting in the field of open source development. I would like to know if i can contribute to this, since it' a good first issue. Thank you in advance.

You can definately contirbute. Are you familiar with python and xarray?

**Comment by GAMinsect** [comment]
I'm familiar with python and common libraries like pandas, but not with xarray

**Comment by peterdudfield** [comment]
thanks good @GAMinsect, you migh need to learn a bit of xarray. Your welcome to give it ago.

**Comment by GAMinsect** [comment]
After 2 weeks of working on it in my free time, here's my implementation.
I formatted the code following your coding style.

```python
def _dataQualityFilter(ds: xr.Dataset) -> bool:
    """Filter out data that is not of sufficient quality."""
    if ds == xr.Dataset():
        return False

    zeroCount = 0
    elementCount = 0
    # Carry out a basic data quality check
    for data_var in ds.data_vars:
        if ds[f"{data_var}"].isnull().any():
            log.warn(
                event=f"Dataset has NaNs in variable {data_var}",
                initTime=str(ds.coords["init_time"].values[0])[:16],
                variable=data_var,
            )

        data = ds[data_var].data
        elementCount += data.size
        zeroCount += (data == 0).sum()

    if zeroCount / elementCount > 0.2:
        raise ValueError("In your dataset more than 20% of your data are 0's")

    return True

**Comment by GAMinsect** [comment]
@peterdudfield is the code fine?

**Comment by peterdudfield** [comment]
ill let @devsjc review if thats ok.

**Comment by GAMinsect** [comment]
@devsjc how's the review going?

**Comment by devsjc** [comment]
Hi @GAMinsect, thanks for your work looking into this, and apologies for the slow response. I'm afraid I've just merged in a quite comprehensive restructuring of the project in order to improve the speed of the application, which has resulted in the `_dataQualityFilter` function being removed, in favour of a new, in progress, `PostProcess` API that I'm in the process of finalising. When complete, I suspect this will be where this check will end up.

Your work wasn't wasted though, as I'll be able to use the logic you've figured out here in that, once it's set up!

**Comment by GAMinsect** [comment]
@devsjc Thank you for the reply (and sorry for my late one) i guess this issue can be closed now, right?

**Comment by Wendoom-dev** [comment]
Hi! If this issue is still open I would like to contribute. @peterdudfield

**Comment by peterdudfield** [comment]
Any update on the @devsjc or is this done already

**Comment by musharafmaqbool** [comment]
ok do let me know once you are sure whether it's done or not?
